File: client/client-multi/client-java-core/src/test/java/org/kaaproject/kaa/client/logging/MemoryLogStorageTest.java
/*
 * Copyright 2014 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.logging;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;

import org.junit.Assert;
import org.junit.Test;

public class MemoryLogStorageTest {

    private void testAddHelper(int addedN, int blockSize, int expectedN) {
        MemoryLogStorage storage = new MemoryLogStorage(blockSize);
        List<LogRecord> expectedList = new LinkedList<>();
        LogRecord record = new LogRecord();

        while (addedN-- > 0) {
            storage.addLogRecord(record);
        }

        while (expectedN-- > 0) {
            expectedList.add(record);
        }

        LogBlock group = storage.getRecordBlock(blockSize);
        List<LogRecord> actualList = group.getRecords();

        Assert.assertTrue("Expected: " + expectedList.size() + ", actual: " + actualList.size()
                , expectedList.size() == actualList.size());

        Iterator<LogRecord> expectedIt = expectedList.iterator();
        Iterator<LogRecord> actualIt = actualList.iterator();

        while (expectedIt.hasNext()) {
            LogRecord expected = expectedIt.next();
            LogRecord actual = actualIt.next();

            Assert.assertTrue(expected.getSize() == actual.getSize());
            Assert.assertArrayEquals(expected.getData(), actual.getData());
        }
    }

    @Test
    public void testEmptyLogRecord() {
        long bucketSize = 3;
        MemoryLogStorage storage = new MemoryLogStorage(bucketSize);
        LogBlock group = storage.getRecordBlock(5);
        Assert.assertTrue(group == null);
    }

    @Test
    public void testRecordCountAndConsumedBytes() {
        long bucketSize = 3;
        MemoryLogStorage storage = new MemoryLogStorage(bucketSize);
        LogRecord record = new LogRecord();
        int insertionCount = 3;

        /*
         * Size of each record is 3B
         */
        int iter = insertionCount;
        while (iter-- > 0) {
            storage.addLogRecord(record);
        }

        Assert.assertTrue(storage.getRecordCount() == insertionCount);
        Assert.assertTrue(storage.getConsumedVolume() == (insertionCount * record.getSize()));
    }

    @Test
    public void testUniqueIdGeneration() {
        long bucketSize = 3;
        MemoryLogStorage storage = new MemoryLogStorage(bucketSize);
        LogRecord record = new LogRecord();

        int insertionCount = 3;

        /*
         * Size of each record is 3B
         */
        int iter = insertionCount;
        while (iter-- > 0) {
            storage.addLogRecord(record);
        }

        LogBlock group1 = storage.getRecordBlock(6);
        LogBlock group2 = storage.getRecordBlock(6);

        Assert.assertNotEquals(group1.getBlockId(), group2.getBlockId());
    }

    @Test
    public void testLogRecordAdding() {
        /*
         * Size of each record is 3B
         */
        testAddHelper(1, 3, 1);
        testAddHelper(4, 3, 1);
        testAddHelper(3, 9, 3);
        testAddHelper(5, 5, 1);
    }

    @Test
    public void testGetSameLogBlock() {
        long bucketSize = 3;
        MemoryLogStorage storage = new MemoryLogStorage(bucketSize);
        LogRecord record = new LogRecord();

        int insertionCount = 3;

        /*
         * Size of each record is 3B
         */
        int iter = insertionCount;
        while (iter-- > 0) {
            storage.addLogRecord(record);
        }

        LogBlock group1 = storage.getRecordBlock(7);

        storage.notifyUploadFailed(group1.getBlockId());

        LogBlock group2 = storage.getRecordBlock(7);

        Assert.assertTrue("Expected: " + group1.getRecords().size() + ", actual: " + group2.getRecords().size()
                , group1.getRecords().size() == group2.getRecords().size());

        Iterator<LogRecord> expectedIt = group1.getRecords().iterator();
        Iterator<LogRecord> actualIt = group2.getRecords().iterator();

        while (expectedIt.hasNext()) {
            LogRecord expected = expectedIt.next();
            LogRecord actual = actualIt.next();

            Assert.assertTrue(expected.getSize() == actual.getSize());
            Assert.assertArrayEquals(expected.getData(), actual.getData());
        }
    }

    @Test
    public void testLogRecordRemoval() {
        long bucketSize = 3;
        MemoryLogStorage storage = new MemoryLogStorage(bucketSize);
        LogRecord record = new LogRecord();

        int insertionCount = 7;

        /*
         * Size of each record is 3B
         */
        int iter = insertionCount;
        while (iter-- > 0) {
            storage.addLogRecord(record);
        }

        LogBlock removingBlock = storage.getRecordBlock(7);

        insertionCount -= removingBlock.getRecords().size();
        storage.removeRecordBlock(removingBlock.getBlockId());

        removingBlock = storage.getRecordBlock(9);

        insertionCount -= removingBlock.getRecords().size();
        storage.removeRecordBlock(removingBlock.getBlockId());

        LogBlock leftBlock = storage.getRecordBlock(50);
        Assert.assertTrue(leftBlock.getRecords().size() == insertionCount);
    }

    @Test
    public void testComplexLogRemoval() {
        long bucketSize = 3;
        MemoryLogStorage storage = new MemoryLogStorage(bucketSize);
        LogRecord record = new LogRecord();

        int insertionCount = 8;

        /*
         * Size of each record is 3B
         */
        int iter = insertionCount;
        while (iter-- > 0) {
            storage.addLogRecord(record);
        }

        LogBlock removingBlock1 = storage.getRecordBlock(7);
        insertionCount -= removingBlock1.getRecords().size();

        LogBlock removingBlock2 = storage.getRecordBlock(9);
        insertionCount -= removingBlock2.getRecords().size();

        LogBlock removingBlock3 = storage.getRecordBlock(6);
        insertionCount -= removingBlock3.getRecords().size();

        storage.removeRecordBlock(removingBlock2.getBlockId());
        storage.notifyUploadFailed(removingBlock1.getBlockId());
        insertionCount += removingBlock1.getRecords().size();

        LogBlock leftBlock = storage.getRecordBlock(50);
        Assert.assertTrue("Ac: " + leftBlock.getRecords().size() + ", ex: " + insertionCount
                , leftBlock.getRecords().size() == insertionCount);
    }

    @Test
    public void testOldestRecordRemoval() {
        long bucketSize = 3;
        MemoryLogStorage storage = new MemoryLogStorage(bucketSize);
        LogRecord record = new LogRecord();

        int insertionCount = 8;

        /*
         * Size of each record is 3B
         */
        int iter = insertionCount;
        while (iter-- > 0) {
            storage.addLogRecord(record);
        }

        long maxAllowedSize = insertionCount * record.getSize() - 4;
        storage.removeOldestRecord(maxAllowedSize);

        LogBlock block = storage.getRecordBlock(50);

        Assert.assertTrue("Ac: " + block.getRecords().size() + ", ex: " + (maxAllowedSize / record.getSize())
                , block.getRecords().size() == (maxAllowedSize / record.getSize()));
    }
}

File: client/client-multi/client-cpp/test/impl/log/MemoryLogStorageTest.cpp
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <boost/test/unit_test.hpp>

#include <string>
#include <cmath>

#include "kaa/log/LogRecord.hpp"
#include "kaa/log/MemoryLogStorage.hpp"
#include "kaa/common/exception/KaaException.hpp"
#include "kaa/log/LogStorageConstants.hpp"
#include "kaa/context/SimpleExecutorContext.hpp"
#include "kaa/KaaClientContext.hpp"
#include "kaa/KaaClientProperties.hpp"
#include "kaa/logging/DefaultLogger.hpp"

#include "headers/MockKaaClientStateStorage.hpp"
#include "headers/context/MockExecutorContext.hpp"

namespace kaa {

static KaaClientProperties properties;
static DefaultLogger tmp_logger(properties.getClientId());
static IKaaClientStateStoragePtr tmp_state(new MockKaaClientStateStorage);
static MockExecutorContext tmpExecContext;
static KaaClientContext clientContext(properties, tmp_logger, tmpExecContext, tmp_state);

#define LOG_TEST_DATA "test data"

static std::int32_t mockBlocksCount = 10000;

static LogRecord createSerializedLogRecord()
{
    KaaUserLogRecord logRecord;
    logRecord.logdata = LOG_TEST_DATA;

    return LogRecord(logRecord);
}

BOOST_AUTO_TEST_SUITE(MemoryLogStorageTestSuite)

BOOST_AUTO_TEST_CASE(BadInitializationParamsTest)
{
    BOOST_CHECK_THROW(
            {
                MemoryLogStorage logStorage(clientContext, 100500, (float)-1.0);
            }, KaaException);

    BOOST_CHECK_THROW(
            {
                MemoryLogStorage logStorage(clientContext, 100500, (float)100.1);
            }, KaaException);
}

BOOST_AUTO_TEST_CASE(BucketizeIsLessThanLogRecordSizeTest)
{
    auto serializedLogRecord = createSerializedLogRecord();

    MemoryLogStorage logStorage(clientContext, serializedLogRecord.getSize() / 2);

    BOOST_CHECK_THROW(logStorage.addLogRecord(std::move(serializedLogRecord)), KaaException);
}

BOOST_AUTO_TEST_CASE(AddRecordsAndCheckStatusTest)
{
    MemoryLogStorage logStorage(clientContext);
    std::size_t serializedLogSize = createSerializedLogRecord().getSize();

    std::srand(std::time(nullptr));
    /*
     * At least 1 records.
     */
    std::size_t logRecordCount = 1 + rand() % 10;

    for (std::size_t i = 1; i <= logRecordCount; ++i) {
        logStorage.addLogRecord(createSerializedLogRecord());
    }

    BOOST_CHECK_EQUAL(logStorage.getStatus().getRecordsCount(), logRecordCount);
    BOOST_CHECK_EQUAL(logStorage.getStatus().getConsumedVolume(), logRecordCount * serializedLogSize);
}

BOOST_AUTO_TEST_CASE(BucketSizeInRecordsConstraint)
{
    const size_t recordInBucket = 2;
    MemoryLogStorage logStorage(clientContext, LogStorageConstants::DEFAULT_MAX_BUCKET_SIZE, recordInBucket);

    /*
     * At least 1 records.
     */
    std::size_t logRecordCount = 1 + rand() % 10;

    for (std::size_t i = 1; i <= logRecordCount; ++i) {
        logStorage.addLogRecord(createSerializedLogRecord());
    }

    for (std::size_t i = 0; i < std::ceil((float)logRecordCount / recordInBucket); ++i) {
        LogBucket bucket = logStorage.getNextBucket();
        BOOST_CHECK(!bucket.getRecords().empty());
    }

    LogBucket bucket = logStorage.getNextBucket();
    BOOST_CHECK(bucket.getRecords().empty());
}

BOOST_AUTO_TEST_CASE(BucketSizeInBytesConstraint)
{
    std::srand(std::time(nullptr));

    std::size_t serializedLogSize = createSerializedLogRecord().getSize();
    std::size_t bucketSizeInBytes = serializedLogSize * (1 + rand() % 2);
    std::size_t recordsInBucket = bucketSizeInBytes / serializedLogSize;
    MemoryLogStorage logStorage(clientContext, bucketSizeInBytes, LogStorageConstants::DEFAULT_MAX_BUCKET_RECORD_COUNT);

    /*
     * At least 1 records.
     */
    std::size_t logRecordCount = 1 + rand() % 10;

    for (std::size_t i = 1; i <= logRecordCount; ++i) {
        logStorage.addLogRecord(createSerializedLogRecord());
    }

    for (std::size_t i = 0; i < std::ceil((float)logRecordCount / recordsInBucket); ++i) {
        LogBucket bucket = logStorage.getNextBucket();
        BOOST_CHECK(!bucket.getRecords().empty());
    }

    LogBucket bucket = logStorage.getNextBucket();
    BOOST_CHECK(bucket.getRecords().empty());
}

BOOST_AUTO_TEST_CASE(GetStatusAfterLogBlockTest)
{
    /*
     * At least 2 records.
     */
    std::size_t logRecordCount = 2 + rand() % 10;
    std::size_t serializedLogSize = createSerializedLogRecord().getSize();
    std::size_t recordBucketSize = (logRecordCount * serializedLogSize) / 2;

    MemoryLogStorage logStorage(clientContext, recordBucketSize);

    for (std::size_t i = 1; i <= logRecordCount; ++i) {
        logStorage.addLogRecord(createSerializedLogRecord());
    }

    auto bucket = logStorage.getNextBucket();

    BOOST_CHECK_EQUAL(logStorage.getStatus().getRecordsCount(), (logRecordCount - bucket.getRecords().size()));
    BOOST_CHECK_EQUAL(logStorage.getStatus().getConsumedVolume(), ((logRecordCount - bucket.getRecords().size()) * serializedLogSize));
}

BOOST_AUTO_TEST_CASE(RemoveLogBlockAndGetStatusTest)
{
    /*
     * At least 2 records.
     */
    std::size_t logRecordCount = 2 + rand() % 10;
    std::size_t serializedLogSize = createSerializedLogRecord().getSize();
    std::size_t recordBucketSize1 = (logRecordCount * serializedLogSize) / 2;

    MemoryLogStorage logStorage(clientContext, recordBucketSize1);
    for (std::size_t i = 1; i <= logRecordCount; ++i) {
        logStorage.addLogRecord(createSerializedLogRecord());
    }

    auto bucket1 = logStorage.getNextBucket();

    logStorage.removeBucket(bucket1.getBucketId());

    BOOST_CHECK_EQUAL(logStorage.getStatus().getRecordsCount(), (logRecordCount - bucket1.getRecords().size()));
    BOOST_CHECK_EQUAL(logStorage.getStatus().getConsumedVolume(),
                     ((logRecordCount - bucket1.getRecords().size()) * serializedLogSize));

    auto bucket2 = logStorage.getNextBucket();

    logStorage.removeBucket(bucket2.getBucketId());

    BOOST_CHECK_EQUAL(logStorage.getStatus().getRecordsCount(), logRecordCount - (recordBucketSize1 / serializedLogSize * 2));
    BOOST_CHECK_EQUAL(logStorage.getStatus().getConsumedVolume(), (logRecordCount * serializedLogSize - (recordBucketSize1 / serializedLogSize * 2) * serializedLogSize));
}

BOOST_AUTO_TEST_CASE(NotifyUploadFailedAndGetStatusTest)
{
    /*
     * At least 2 records.
     */
    std::size_t logRecordCount = 2 + rand() % 10;
    std::size_t serializedLogSize = createSerializedLogRecord().getSize();
    std::size_t recordBucketSize1 = (logRecordCount * serializedLogSize) / 2;

    MemoryLogStorage logStorage(clientContext, recordBucketSize1, (std::size_t)mockBlocksCount);

    for (std::size_t i = 1; i <= logRecordCount; ++i) {
        logStorage.addLogRecord(createSerializedLogRecord());
    }

    auto bucket1 = logStorage.getNextBucket();

    BOOST_CHECK_EQUAL(logStorage.getStatus().getRecordsCount(), (logRecordCount - bucket1.getRecords().size()));
    BOOST_CHECK_EQUAL(logStorage.getStatus().getConsumedVolume(),
                     ((logRecordCount - bucket1.getRecords().size()) * serializedLogSize));

    logStorage.rollbackBucket(bucket1.getBucketId());

    BOOST_CHECK_EQUAL(logStorage.getStatus().getRecordsCount(), logRecordCount);
    BOOST_CHECK_EQUAL(logStorage.getStatus().getConsumedVolume(), (logRecordCount * serializedLogSize));
}

BOOST_AUTO_TEST_CASE(ForceRemovalOfAllLogsTest)
{
    std::size_t logRecordCount = 5;
    std::size_t serializedLogSize = createSerializedLogRecord().getSize();
    std::size_t maxLogStorageSize = logRecordCount * serializedLogSize;
    float percentToDelete = 100.0;

    MemoryLogStorage logStorage(clientContext, maxLogStorageSize, percentToDelete);
    for (std::size_t i = 1; i <= logRecordCount; ++i) {
        logStorage.addLogRecord(createSerializedLogRecord());
    }

    /*
     * Should cause force removal
     */
    logStorage.addLogRecord(createSerializedLogRecord());

    BOOST_CHECK_EQUAL(logStorage.getStatus().getRecordsCount(), 1);
    BOOST_CHECK_EQUAL(logStorage.getStatus().getConsumedVolume(), serializedLogSize);
}

BOOST_AUTO_TEST_CASE(ForceRemovalOfSpecifiedPercentOfLogsTest)
{
    std::size_t logRecordCount = 10;
    std::size_t serializedLogSize = createSerializedLogRecord().getSize();
    std::size_t maxLogStorageSize = logRecordCount * serializedLogSize;
    float percentToDelete = 51.1;

    MemoryLogStorage logStorage(clientContext, maxLogStorageSize, percentToDelete, LogStorageConstants::DEFAULT_MAX_BUCKET_SIZE, logRecordCount / 2);

    for (std::size_t i = 1; i <= logRecordCount; ++i) {
        logStorage.addLogRecord(createSerializedLogRecord());
    }

    logStorage.getNextBucket();

    /*
     * Should cause force removal
     */
    logStorage.addLogRecord(createSerializedLogRecord());

    std::size_t recordLeftAfterRemoval = 1 + ((maxLogStorageSize * (100.0 - percentToDelete) / 100.0) / serializedLogSize);
    size_t sizeAfterRemoval = recordLeftAfterRemoval * serializedLogSize;

    BOOST_CHECK_EQUAL(logStorage.getStatus().getRecordsCount(), recordLeftAfterRemoval);
    BOOST_CHECK_EQUAL(logStorage.getStatus().getConsumedVolume(), sizeAfterRemoval);
}

BOOST_AUTO_TEST_SUITE_END()

}
